scipy/optimize/_trustregion_exact.py

```diff
@@ -0,0 +1,425 @@
+
+            # Count iteration
+            self.niter += 1
+
+            # Check if factorization succeded
+            if info == 0 and self.jac_mag > self.CLOSE_TO_ZERO:
+                # Successfull factorization
+
+                # Solve `U.T s = -g`
+                s = solve_triangular(U, -1*self.jac, trans='T')
+
+                # Solve `U p = s`
+                p = solve_triangular(U, s, trans='N')
+
+                # Compute `p` norm
+                pnorm = norm(p)
+
+                # Check for interior convergence
+                if pnorm <= trust_radius and damp == 0:
+                    hits_boundary = False
+                    break 
+
+                # Solve `U.T w = p`
+                w = solve_triangular(U, p, trans='T')
+
+                # Compute `w` norm
+                wnorm = norm(w)
+
+                # Compute Newton step accordingly to
+                # formula (4.44) p.87 from ref [2]
+                delta_damp = (pnorm/wnorm)**2*(pnorm-trust_radius)/trust_radius
+                damp_next = damp + delta_damp
+
+                if pnorm < trust_radius:  # Inside boundary
+                    smin, zmin = estimate_smallest_singular_value(U)
+
+                    # Compute `steplen` such that
+                    # norm(p+steplen*zmin)=trust_radius
+                    # minimize the model magnitude.
+                    # Use formula from page 6 from ref [3]
+                    pdotz = np.dot(p, zmin)
+                    num = trust_radius**2-pnorm**2
+                    den = pdotz+np.sign(pdotz)*np.sqrt(pdotz**2+num)
+                    steplen = num/den
+
+                    # Compute the quadratic term  (p.T*H*p)
+                    quadratic_term = np.dot(p, np.dot(H, p))
+
+                    # Check stop criteria
+                    relative_error = steplen**2*smin**2/(quadratic_term +
+                                                         damp*trust_radius**2)
+                    if relative_error <= self.k_hard:
+                        p = p+steplen*zmin
+                        break
+
+                    # Update uncertanty bounds
+                    ub = damp
+                    lb = max(lb, damp-smin**2)
+
+                    # Compute Cholesky factorization
+                    H = self.hess+damp_next*np.eye(n)
+                    c, info = self.cholesky(H, lower=False,
+                                            overwrite_a=False,
+                                            clean=True)
+
+                    # Check if the factorization have succeded
+                    #
+                    if info == 0:  # Successfull factorization
+                        # Update damping factor
+                        damp = damp_next
+                        already_factorized = True
+                    else:  # Unsuccessfull factorization
+                        # Update uncertanty interval
+                        lb = max(lb, damp_next)
+
```

Comment:
The listing in the books says to check "lamba^L for interior convergence" (step 5b). I guess it isn't necessary, and the authors weren't thinking in terms of actual programming. But please check this points.